### ISSUE TITLE
Perform erl_lint checks when building the AST

### ISSUE THREAD

**Issue by whatyouhide** [elixir-lang/elixir#5800]
We aim to perform as many checks that `erl_lint` performs as we can when building the Elixir AST (during expansion). This issue exists to keep track of the checks that `erl_lint` performs; some of them we already perform, some of them are invalid for Elixir, and some of them we want to implement. [Reference](https://github.com/erlang/otp/blob/bd25898d3bf63316fcf116ba03f92c12500927fd/lib/stdlib/src/erl_lint.erl#L151-L407).

## General

- [x] `undefined_module`: no module definition
- [x] `redefine_module`
- [x] `pmod_unsupported`: parametrized modules are no longer supported
- [x] `{redefine_mod_import, M, P}`: module already imported from package `P`
- [x] `invalid_call`: invalid function call
- [x] `invalid_record`: invalid record expression
- [x] `{attribute,A}`: attribute after function definitions
- [x] `{missing_qlc_hrl,A}`: `qlc:q` called but `qlc.hrl` not included
- [x] `{redefine_import,{{F,A},M}}`: function already imported from another module
- [x] `{bad_inline,{F,A}}`: inlined function undefined
- [x] `{invalid_deprecated,D}`: badly formed deprecated attribute
- [x] `{bad_deprecated,{F,A}}`: depreacted function undefined or not exported
- [x] `{bad_nowarn_unused_function,{F,A}}`: function undefined
- [x] `{bad_nowarn_bif_clash,{F,A}}`: function undefined
- [x] `disallowed_nowarn_bif_clash`: `-compile` directive `nowarn_bif_clash` is no longer allowed
- [x] `{bad_nowarn_deprecated_function,{M,F,A}}`: not a depreacted function
- [x] `{bad_on_load,Term}`: badly formed `on_load` attribute
- [x] `multiple_on_loads`: multiple `on_load` attributes
- [x] `{bad_on_load_arity,{F,A}}`: arity must be `0`
- [x] `{undefined_on_load,{F,A}}`: function for `on_load` undefined
- [x] `export_all`: `export_all` flag enabled
- [x] `{duplicated_export, {F,A}}`: function already exported
- [x] `{unused_import,{{F,A},M}}`
- [x] `{undefined_function,{F,A}}`
- [x] `{redefine_function,{F,A}}`: function already defined
- [x] `{define_import,{F,A}}`: defining imported function
- [x] `{unused_function,{F,A}}`
- [x] `{call_to_redefined_bif,{F,A}}`: ambiguous call of overridden auto-imported BIF, use `erlang:...` or `-compile({no_auto_import, [f/a]})`
- [x] `{call_to_redefined_old_bif,{F,A}}`: ambiguous call of overridden pre R14 auto-imoported BIF
- [x] `{redefine_old_bif_import,{F,A}}`: import overrides pre R14 auto-imported BIF, use `no_auto_import` compile directive
- [x] `{redefine_bif_import,{F,A}}`: import overrides auto-imported BIF
- [x] `{deprecated, MFA, ReplacementMFA, Rel}`: `MFA` will be deprecated in release `Rel`
- [x] `{deprecated, {M1, F1, A1}, String}`
- [x] `{deprecated_type, {M1, F1, A1}, String}`
- [x] `{removed, MFA, ReplacementMFA, Rel}`: removed in release `Rel`
- [x] `{removed, MFA, String}`
- [x] `{removed_type, MNA, ReplacementMNA, Rel}`: removed in release `Rel`
- [x] `{obsolete_guard, {F, A}}`: use `is_F`
- [x] `{obsolete_guard_overridden,Test}`
- [x] `{too_many_arguments,Arity}`: max allowed 255

## Patterns and guards

- [x] `illegal_pattern`
- [x] `illegal_pattern` for `+x` and `-x` in matches (literal numbers are ok though!)
- [x] `illegal_map_key`: illegal map key in pattern
- [x] `illegal_bin_pattern`: binary patterns cannot be matched in parallel using `=`
- [x] `illegal_expr`
- [x] `{illegal_guard_local_call, {F,A}}`: call to local/imported function is illegal in guard
- [x] `illegal_guard_expr`

## Maps

- [x] `illegal_map_construction`: only association operators `=>` are allowed in map construction

## Records

- [x] `{undefined_record,T}`
- [x] `{redefine_record,T}`: already defined
- [x] `{redefine_field,T,F}`: already defined
- [x] `{undefined_field,T,F}`
- [x] `illegal_record_info`
- [x] `{field_name_is_variable,T,F}`: field is not an atom or `_`
- [x] `{wildcard_in_update,T}`: meaningless use of `_` in update of record
- [x] `{unused_record,T}`
- [x] `{untyped_record,T}`: record has field(s) without type info

## Variables

- [x] `{unbound_var,V}`
- [x] `{unsafe_var,V,{What,Where}}`
- [x] `{exported_var,V,{What,Where}}`
- [x] `{shadowed_var,V,In}`
- [x] `{unused_var, V}`
- [x] `{variable_in_record_def,V}`

## Binaries

- [x] `{undefined_bittype,Type}`
- [x] `{bittype_mismatch,Val1,Val2,What}`: conflict in specification for bit field
- [x] `bittype_unit`: bit unit size must not be specified unless a size is specified too
- [x] `illegal_bitsize`
- [x] `unsized_binary_not_at_end`: a binary field without size is only allowed at the end of a binary pattern
- [x] `typed_literal_string`: a literal string in a binary pattern must not have type/size
- [x] `utf_bittype_size_or_unit`: neither size nor unit must be given for segments of type `utf*`
- [x] `{bad_bitsize,Type}`
- [x] `unsized_binary_in_bin_gen_pattern`: binary fields without size are not allowed in patterns of bit string generators

## Behaviours

- [x] `{conflicting_behaviours,{Name,Arity},B,FirstL,FirstB}`: callback required by two behaviours
- [x] `{undefined_behaviour_func, {Func,Arity}, Behaviour}`
- [x] `{undefined_behaviour,Behaviour}`
- [x] `{undefined_behaviour_callbacks,Behaviour}`
- [x] `{ill_defined_behaviour_callbacks,Behaviour}`
- [x] `{ill_defined_optional_callbacks,Behaviour}`
- [x] `{redefine_optional_callback, {F, A}}`: callback is duplicated
- [x] `{undefined_callback, {_M, F, A}}`

## Types and specs

- [x] `{behaviour_info, {_M,F,A}}`: cannot define `callback` attribute when `behaviour_info` is defined
- [x] `{singleton_typevar, Name}`: type variable is used only once
- [x] `{bad_export_type, _ETs}`: bad `export_type`
- [x] `{duplicated_export_type, {T, A}}`
- [x] `{undefined_type, {TypeName, Arity}}`
- [x] `{unused_type, {TypeName, Arity}}`
- [x] `{new_builtin_type, {TypeName, Arity}}`: this is a new builtin type and its (re)definition is allowed only until the next release
- [x] `{builtin_type, {TypeName, Arity}}`: cannot redefine builtin type
- [x] `{renamed_type, OldName, NewName}`: type has been renamed
- [x] `{redefine_type, {TypeName, Arity}}`
- [x] `{type_syntax, Constr}`: bad type
- [x] `old_abstract_code`: abstract code generated before Erlang/OTP 19.0 and having typed record fields cannot be compiled
- [x] `{redefine_spec, {M, F, A}}`
- [x] `{redefine_spec, {F, A}}`
- [x] `{redefine_callback, {F, A}}`
- [x] `{bad_callback, {M, F, A}}`: explicit module not allowed for callback
- [x] `{spec_fun_undefined, {F, A}}`: spec for undefined function
- [x] `{missing_spec, {F,A}}`: missing spec for function
- [x] `spec_wrong_arity`: spec has wrong arity
- [x] `callback_wrong_arity`: callback has wrong arity
- [x] `{deprecated_builtin_type, {Name, Arity}, Replacement, Rel}`
- [x] `{not_exported_opaque, {TypeName, Arity}}`: opaque type is not exported
- [x] `{underspecified_opaque, {TypeName, Arity}}`: opaque type is underspecified and therefore meaningless

**Comment by josevalim** [comment]
Note that we can only effectively disable erl_lint once we depend on Erlang 20 exclusively.

**Comment by michalmuskala** [comment]
Should we move the `module_info` check to be handled similar to other lint checks we're doing here? I think it would be more consistent.

**Comment by josevalim** [comment]
@michalmuskala please expand. :)

**Comment by michalmuskala** [comment]
It's invalid for user to define a `module_info/0,1` function, since those are generated by the erlang compiler. Today the check is done by the implementation of the `def` macro, but I think it would be better to group it with other linting checks that we're defining here.

**Comment by josevalim** [comment]
I don't think there is such group though. The checks here are being applied to different parts of the codebase, depending on the part they are relevant to.

**Comment by josevalim** [comment]
@fertapric good catch! We may have ticked the wrong one off. `renamed_type` is irrelevant to Elixir and `redefine_type` should indeed be done in Elixir land indeed.

**Comment by fertapric** [comment]
I think `{singleton_typevar, Name}` does not apply to Elixir. 

I believe this `erl_lint` check only applies in cases in which a variable is used only once in a spec definition. For example:

```erlang
-spec foo(X, integer) -> boolean
```

However, Elixir only defines variables when guards are used:

```elixir
@spec foo(x, integer) :: boolean when x: integer
```

If there are not guards, Elixir cannot know if the given identifier is a type or a variable, and it always parses them as types. For example, these two examples always raise `CompileError: type x() undefined`:

```elixir
@spec foo(x, x) :: boolean
@spec foo(x, integer) :: boolean
```

That's why the `singleton_typevar` scenario cannot happen.

**Comment by josevalim** [comment]
@hat happens if you do `@spec foo(x, integer) :: boolean when x: var` though, which would be the precise equivalent to erlang type vars?

**Comment by fertapric** [comment]
@josevalim Erlang does not raise in those cases, for example:

```erlang
-spec foo(X, integer) -> boolean when X :: integer.
```

**Comment by josevalim** [comment]
@fertapric note in my original comment that the type at the end needs to be `var` though, not `integer`. That's because variables in elixir types are aliases, not variables, so we need to tag them.

In other words, `foo(X, integer) -> boolean` is `foo(x, integer) :: integer when x: var` in Elixir.

**Comment by fertapric** [comment]
Oh! I see.

While reading the documentation, I thought `var` was a custom type. Should `var` be restricted then?

For example, this works:

```elixir
defmodule RedefineVarKeyword do
  @type var :: atom

  @spec foo(x, integer) :: boolean when x: var
  def foo(x, y), do: x == y
end
```

In addition, should the typespecs documentation be updated to say that `var` is a special keyword? For example:

> Type variables with no restriction can also be defined by tagging them with `var`.
>
>     @spec function(arg) :: [arg] when arg: var

**Comment by josevalim** [comment]
> For example, this compiles but the var type is not applied in the guard (and it might be confusing):

I think we should warn if you use `var` anywhere except the `name: var` notation. We should also leave a TODO to raise in the future. For all purposes, `var` is treated as a special type.

**Comment by josevalim** [comment]
Hrm, on the other hand, `@spec foo(x, x, var) :: boolean when x: var` is not ambiguous to the language. So maybe let's hold back on the warning. We can always add it later.

**Comment by fertapric** [comment]
I was thinking on this use case:

```elixir
defmodule DefineVarType do
  @type my_var :: integer
  @type var :: integer

  @spec foo(x, x, var) :: boolean when x: var
  def foo(x, y, z), do: x + y == z

  @spec bar(x, x, my_var) :: boolean when x: my_var
  def bar(x, y, z), do: x + y == z
end
```

If I got it right, `bar/3` arguments can only be `integer`, and a developer might expect `foo/3` to behave the same.

**Comment by josevalim** [comment]
So I think we need to declare `var/0` as a built-in type and not allow `@type var` in the first place. We will have to change the usage in `Macro.Env.var` too.

**Comment by fertapric** [comment]
@josevalim I think we can tick "`{renamed_type, OldName, NewName}`: type has been renamed" for the reasons explained [in this comment](https://github.com/elixir-lang/elixir/issues/5800#issuecomment-408677585)

I'm currently working on "`{singleton_typevar, Name}`: type variable is used only once" but I will tackle the declaration of `var/0` as a built-in type in a different PR.

**Comment by josevalim** [comment]
Thanks everyone for the amazing work! We can already reap the benefits of this in #8241, where the compiler got 5-10% faster. There is only `{type_syntax, _}` left and I will open up a separate issue to track it. :heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
